extension/jaegerremotesampling/provider.go

```diff
@@ -0,0 +1,381 @@
+// Copyright (c) 2018 The Jaeger Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package jaegerremotesampling
+
+import (
+	"bytes"
+	"context"
+	"encoding/gob"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"sync/atomic"
+	"time"
+
+	"github.com/jaegertracing/jaeger/proto-gen/api_v2"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/extension/jaegerremotesampling/internal"
+)
+
+// null represents "null" JSON value and
+// it un-marshals to nil pointer.
+var nullJSON = []byte("null")
+
+type samplingProvider struct {
+	logger *zap.Logger
+
+	storedStrategies atomic.Value // holds *storedStrategies
+
+	cancelFunc context.CancelFunc
+
+	options Options
+}
+
+type storedStrategies struct {
+	defaultStrategy   *api_v2.SamplingStrategyResponse
+	serviceStrategies map[string]*api_v2.SamplingStrategyResponse
+}
+
+type strategyLoader func() ([]byte, error)
+
+// newProvider creates a strategy store that holds static sampling strategies.
+func newProvider(options Options, logger *zap.Logger) (internal.Provider, error) {
```

Comment:
internal.Provider is an interface. You have two possible implementations, File and Remote. Calling this function `newProvider` when it's actually creating FileProvider is misleading

Reply:
also, the main Config calls this `Source`, so I would get rid of "provider" nomenclature altogether and have FileSource & RemoteSource (both should be in internal package)

Reply:
That makes sense, I'll try and improve it

Reply:
Still need to clean up a bit before updating the PR, but here's the new `internal` package structure:

```
└── internal
    ├── mocks
    │   └── mock_source.go (Needed for grpc_test and http_test)
    ├── server
    │   ├── grpc
    │   │   ├── grpc.go
    │   │   ├── grpc_handler.go
    │   │   ├── grpc_handler_test.go
    │   │   └── grpc_test.go
    │   └── http
    │       ├── http.go
    │       └── http_test.go
    └── source
        ├── interface.go
        ├── filesource
        │   ├── constants.go
        │   ├── fixtures/ (used for testing)
        │   ├── model.go
        │   ├── options.go
        │   ├── filesource.go (previously provider.go; newProvider will be renamed NewFileSource function that returns a source.Source)
        │   └── filesource_test.go
        └── remotesource
            ├── manager.go
            ├── manager_test.go
            ├── remote_strategy_cache.go
            ├── remote_strategy_cache_test.go
            └── remote_strategy_store.go (rename NewRemoteStrategyStore to NewRemoteSource, returns a source.Source)
```